### PR TITLE
Polkadot gets topic-based message IDs

### PR DESCRIPTION
**Pull request by gavofyork**
Integrating https://github.com/paritytech/polkadot/pull/7234 into Polkadot.

Do not merge until after audit.

### CODE REVIEW

**Review by bkontur — APPROVED** [review]

File: xcm/src/v3/junction.rs [hunk 1]
```diff
@@ -231,6 +231,7 @@ pub enum Junction {
 	/// An instanced, indexed pallet that forms a constituent part of the context.
 	///
 	/// Generally used when the context is a Frame-based chain.
+	// TODO XCMv4 inner should be `Compact<u32>`.
```
Comment (KiChjang) [inline xcm/src/v3/junction.rs:234]: Can we ever have a runtime that has over 256 pallets? If not, then I'm not sure why it has to be u32 then.

**Review by KiChjang — APPROVED** [review]

File: xcm/src/v3/junction.rs (hunk 1 above)
Comment (gavofyork) [reply xcm/src/v3/junction.rs:234]: "ever" is a long time, and there's no great reason why there should not be more than 256 pallet instances.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3371624

**Review by ordian — APPROVED** [review]

**Comment by bkontur** [comment]
bot merge

**Pull request merged**